docs: add message concepts doc (#944)
asyncapi/website: pages/docs/concepts/message.md

---
title: Message
weight: 5
---

## What is a message?

> A `message` is a method of communication used to transmit or exchange information from a sender to the receiver through the use of `channels`. A single `message` can be consumed by multiple independent receivers, and can be also defined as an _event_, _command_, or _request_. The sender includes a payload containing data that needs to be processed and serialized into a format appropriate for the receiver to process.

Reviewer: ```suggestion
A `message` is a method of communication used to transmit or exchange information from a sender to the receiver through the use of `channels`. A single `message` can be consumed by multiple independent receivers, and can also be defined as an _event_ or _command_. The sender includes a payload containing data that needs to be processed and serialized into a format appropriate for the receiver to process.
```

I suggested to remove `request` in the other comment, maybe you missed it

> The sender includes a payload containing data that needs to be processed and serialized into a format appropriate for the receiver to process

regarding this sentence, I still have doubts

I read `The sender includes a payload containing data`

any immediately, my brain asks questions:
- sender includes payload where?
- payload containing data? but payload is data, 

also `data that needs to be processed and serialized` sounds like it is not the receiver that needs to do it, but it is receiver that is doing it

maybe try to explain if new sentences what you want to explain so we can come up with something. Maybe @alequetzalli can help coming up with some ideas

> A `message` is a method of communication used to transmit or exchange information from a sender to the receiver through the use of `channels`. A single `message` can be consumed by multiple independent receivers, and can also be defined as an _event_ or _command_. The sender includes in the message a payload of data that needs to be processed and serialized by the receiver into an appropriate format e.g JSON, XML, binary, etc.

Author: @derberg how is it sounding now?
cc. @alequetzalli

``` mermaid
graph LR
subgraph one[channel]
b([message])
end

> A{{sender application}} --> b --> C{{receiver application}}

Reviewer: lgtm! I had doubts if we should write `application` but I think it is fine as it corresponds with AsyncAPI goal, that is for applications

---
title: Message

> weight: 2

Reviewer 2: ```suggestion
weight: 5
```

We can't use `2` as your **weight** for this NEW Doc page because your currently merged `server` page already has that weight. Def a good idea to check the latest weight used in the previous pages in each section so that you always know what number to start ordering with. 😄✌🏽

---

## What is a message?

> `Message` is a method of communication used to transmit or exchange information from a sender to the receiver through the use of `channels`. A `message` is generic and can also be defined as an event, command, or request. The sender includes a payload that contains data that would be processed according to the preferred format defined by the receiver.

Reviewer 2: ```suggestion
A `message` is a method of communication used to transmit or exchange information from a sender to the receiver through the use of `channels`. A `message` is generic and can also be defined as an _event_, _command_, or _request_. The sender includes a payload containing data that is processed according to the preferred format defined by the receiver.
```

> `Message` is a method of communication used to transmit or exchange information from a sender to the receiver through the use of `channels`. A `message` is generic and can also be defined as an event, command, or request. The sender includes a payload that contains data that would be processed according to the preferred format defined by the receiver.

Reviewer: ```suggestion
`Message` is a method of communication used to transmit or exchange information from a sender to the receiver through the use of `channels`. A `message` is a generic term and can also be defined as an event or command. The sender includes a payload that contains data that would be processed according to the preferred format defined by the receiver.
```

I suggest we also write that a single message can be consumed by multiple independent receivers.

> A `message` is a generic term

not sure if I'm just a picky non-native speaker, but I think being so detailed is important here. 

> The sender includes a payload that contains data that would be processed according to the preferred format defined by the receiver.

I'm a bit confused about this sentence. Not sure what it means.

Author: So here all i'm saying is that the `sender` includes a payload that contains the information/data which is formatted according to the `receiver's` application e.g json, XML 

Hope that makes sense

Reviewer: the thing is that receiver do not dictate how the message is formatted, so if sender sends json, then receiver much make sure that it can serialize it. This is what I think confuses me here

Author: Oooh yes, i get what you are saying now. Okay let me rephrase it properly then😅

> ## Messages vs Events

Reviewer: from whole twitter conversation about message and events, this is my favorite https://twitter.com/hovenko/status/1569746909186740225 😄 maybe one day we can turn it into a kind of diagram/picture showing Archimedes https://img.freepik.com/darmowe-wektory/archimedes-w-kreskowce-w-wannie-ze-slowem-eureka_1308-100068.jpg?w=2000 as event and then another picture with him where he runs and screams "hey everybody, listen what I've found out" as a message 😆

> A `message` carries information from one application to the other, while an `event` is a message that provides details of something that has already occurred. `Messages` contain all necessary information required to perform the request, while on the other hand, `events` only contain data of the changed state.

Reviewer 2: ```suggestion
A `message` carries information from one application to the other, while an `event` is a message that provides details of something that has already occurred. `Messages` contain all necessary information required to perform the request, while `events` only contain data of the changed state.
```

Overall, `events` are `messages` but not all `messages` are `events`.

``` mermaid
graph LR
subgraph one[message]
b([channel])
end

> A{{sender application}} --> b --> C{{receiver application}}

Reviewer: on the diagram, it looks like the channel is part of the message, while it is a message that goes through the channel. Maybe just revert names? 🤔 

also, please move the diagram under the first section

Author: hey @derberg should I rather do like this then 

![image](https://user-images.githubusercontent.com/66913810/191207686-00968f73-3b65-47bd-ba7a-a8a6365cdac6.png)

Reviewer: no, by revert names I meant that the message is inside a channel, and not channel inside a message

<img width="1747" alt="Screenshot 2022-09-20 at 10 39 44" src="https://user-images.githubusercontent.com/6995927/191210669-c55e0b78-1bb4-40e5-a1d8-7c0cf544a72d.png">

Author: Got it

---
title: Message
weight: 10
---

## What is a message?

> A `message` is a method of communication used to transmit or exchange information from a sender to the receiver through `channels`. A single `message` can be consumed by multiple independent receivers and can also be defined as an _event_ or _command_. The sender includes a payload of data that needs to be processed and serialized by the receiver into an appropriate format, e.g., JSON, XML, binary, etc.

Reviewer 3: ```suggestion
A `message` is a communication asset used to transmit or exchange information from a sender to the receiver through `channels`. A single `message` can be consumed by multiple independent receivers and can also be defined as an _event_ or _command_. The sender includes a payload of data that needs to be processed and serialized by the receiver into an appropriate format, e.g., JSON, XML, binary, etc. It may also include metadata, i.e., information that describes the message itself. This metadata is often known as _headers_ or _properties_.
```

Two things here:
* I suggested _asset_ instead of _method_ because, to me, _method_ kinda implies an action. And the action is actually the operation. Feel free to ignore though 🤷 
* Added info about message metadata. Most systems offer the possibility to attach metadata to the message and it's super useful.

> A `message` is a method of communication used to transmit or exchange information from a sender to the receiver through the use of `channels`. A single `message` can be consumed by multiple independent receivers, and can also be defined as an _event_ or _command_. The sender includes in the message a payload of data that needs to be processed and serialized by the receiver into an appropriate format e.g JSON, XML, binary, etc.

Reviewer 2: ```suggestion
A `message` is a method of communication used to transmit or exchange information from a sender to the receiver through `channels`. A single `message` can be consumed by multiple independent receivers and can also be defined as an _event_ or _command_. The sender includes a payload of data that needs to be processed and serialized by the receiver into an appropriate format, e.g., JSON, XML, binary, etc.
```

``` mermaid
graph LR
subgraph one[channel]
b([message])
end

A{{sender application}} --> b --> C{{receiver application}}
  
```

In the diagram above, the sender application transmits a `message` to the receiver application.

## Messages vs Events

> A `message` carries information from one application to the other, while an `event` is a message that provides details of something that has already occurred. `Messages` contain all necessary information required to perform the request, while `events` only contain data of the changed state.

Reviewer 3: There is a contradiction in this paragraph. If an event is a message and an event is only about a changed state, then we can't assure that messages contain all the necessary information to perform a request. Event messages don't contain any information to perform a request.

`Messages` is just a generic term. They can be `events`, `queries`, or `commands`. And `commands` and `queries` are `requests`. 

Here's a representation of what I mean:

```mermaid
graph TD
    A(Message)
    A --> B(Event)
    A --> C(Request)
    C --> D(Query)
    C --> E(Command)
```

Author: @fmvilas If understand correctly, messages don't necessarily contain information to make a request and events provide information about a changed state. 
So how best can we define/differentiate the two apart from messages being an umbrella term.

Reviewer 3: You can just say that messages contain information but can't anticipate which kind of information. Actually, a message could just contain "Hello Thulie", and that would NOT classify it in any of the described groups (event, query, command). That said, it's not usually the case so we can just simplify and say that, depending on the kind of information the message contains, it may be an event (something that already happened), a query, or a command.

Queries: a query is a request that always expects a response with the results. A query will never ever modify anything on the receiver side. I mean, it's always read-only. It's meant to gather information but never modify it. Example: "give me a list of the last 10 registered users." Notice it doesn't modify anything, just grabs the information and reply.

Command: a command is a request that may or may not receive a response. A command is meant to modify something on the receiver side or trigger some action. Example: "change user email to some@user.com." Notice how it's commanding the receiver to modify data. Depending on the nature of the app, it may or may not reply with the new user information.

Hope that helps :)

Overall, `events` are `messages` but not all `messages` are `events`.

> 

Reviewer 4: Hey @thulieblack, on the completion of this page or the pages you add, please make sure that you add the `Prev` and `UpNext` buttons similar to other pages and update the routes in such a way that it links with the Docs Tree in sidebar.

Author: Sure, will do

---
title: Message
weight: 30
---

## What is a message?
A `message` is a communication asset used to transmit or exchange information from a sender to the receiver through `channels`. A single `message` can be consumed by multiple independent receivers and can also be defined as an _event_ or _command_. The sender includes a payload of data that needs to be processed and serialized by the receiver into an appropriate format, e.g., JSON, XML, binary, etc. It may also include metadata, i.e., information that describes the message itself. This metadata is often known as _headers_ or _properties_.

``` mermaid
graph LR
subgraph one[channel]
b([message])
end

A{{sender application}} --> b --> C{{receiver application}}
  
```

In the diagram above, the sender application transmits a `message` to the receiver application.

## Messages vs Events

> A `message` carries information from one application to the other, while an `event` is a message that provides details of something that has already occurred. One important aspect to note is that depending on the type of information a `message` contains, it can fall under an _event_, _query_, or _command_. 

Author: Hey Fran @fmvilas this is what I came up with, does it make sense or should I revise?

Reviewer 3: You've put it perfectly 👍 👏
